dan_layer/consensus/src/hotstuff/on_propose.rs

```diff
@@ -196,26 +196,33 @@ where TConsensusSpec: ConsensusSpec
         let mut total_leader_fee = 0;
         let commands = batch
             .into_iter()
-            .map(|t| match t.current_stage() {
+            .filter_map(|t| match t.current_stage() {
                 // If the transaction is New, propose to Prepare it
-                TransactionPoolStage::New => Ok(Command::Prepare(t.get_local_transaction_atom())),
+                TransactionPoolStage::New => Some(Ok(Command::Prepare(t.get_local_transaction_atom()))),
                 // The transaction is Prepared, this stage is only _ready_ once we know that all local nodes
                 // accepted Prepared so we propose LocalPrepared
-                TransactionPoolStage::Prepared => Ok(Command::LocalPrepared(t.get_local_transaction_atom())),
+                TransactionPoolStage::Prepared => Some(Ok(Command::LocalPrepared(t.get_local_transaction_atom()))),
                 // The transaction is LocalPrepared, meaning that we know that all foreign and local nodes have
                 // prepared. We can now propose to Accept it. We also propose the decision change which everyone should
                 // agree with if they received the same foreign LocalPrepare.
                 TransactionPoolStage::LocalPrepared => {
-                    let involved = local_committee_shard.count_distinct_buckets(t.transaction().evidence.shards_iter());
-                    let involved = NonZeroU64::new(involved as u64).ok_or_else(|| {
-                        HotStuffError::InvariantError(format!(
-                            "Number of involved shards is zero for transaction {}",
-                            t.transaction_id(),
-                        ))
-                    })?;
-                    let leader_fee = t.calculate_leader_fee(involved, EXHAUST_DIVISOR);
-                    total_leader_fee += leader_fee;
-                    Ok(Command::Accept(t.get_final_transaction_atom(leader_fee)))
+                    if t.transaction().evidence.all_shards_complete() {
```

Comment:
The transaction is never marked as is_ready == true unless we have `evidence.all_shards_complete()`. Only is_ready==true transactions are returned `get_batch_for_next_block`. So this shouldn't be necessary to check again (this change should have no effect). Are you seeing Accept commands before evidence is complete?

Reply:
If there is a bug where is_ready is true, this will create a block with less transactions and hide the bug. There may be a bug in the way we handle work out each transaction atom state from the current chain from the locked block to the leaf block.

Reply:
Or it probably is a race condition, where some validators have seen the foreign local prepared, and some haven't (I think this issue persists). In this case, we'd have to wait for us to get the foreign `LocalPrepared` before voting on the block. I think we abstain from voting in this case currently which is probably correct as long as we're able to recover once the block gets to locked block.